core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1062,45 +1060,62 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
    * in parallel, combining their results with the specified `f` function. If
    * either side fails, then the other side will be interrupted.
    */
-  def zipWithPar[R1 <: R, E1 >: E, A1, A2](that: ZManaged[R1, E1, A1])(f0: (A, A1) => A2): ZManaged[R1, E1, A2] =
+  def zipWithPar[R1 <: R, E1 >: E, A1, A2](that: ZManaged[R1, E1, A1])(f0: (A, A1) => A2): ZManaged[R1, E1, A2] = {
+
+    sealed trait State
+    case object Running   extends State
+    case object Acquiring extends State
+    case object Exited    extends State
+
     ZManaged[R1, E1, A2] {
-      Ref.make[Either[Exit[Any, Any], List[Exit[Any, Any] => ZIO[R1, Nothing, Any]]]](Right(Nil)).map { ref =>
+      ZManaged.FinalizerRef.make[R1](_ => UIO.unit).map { finalizers =>
         Reservation(
           acquire = {
-            val left = ZIO.uninterruptibleMask { restore =>
-              self.reserve.flatMap { reservation =>
-                ref.updateSomeAndGet {
-                  case Right(finalizers) => Right(reservation.release :: finalizers)
-                }.flatMap {
-                  case Left(exit) => restore(reservation.acquire).ensuring(reservation.release(exit))
-                  case _          => restore(reservation.acquire)
-                }
-              }
+            val left = ZIO.uninterruptibleMask {
+              restore =>
+                for {
+                  ref   <- Ref.make[State](Running)
+                  latch <- Promise.make[Nothing, Unit]
+                  res   <- self.reservation
+                  acquire = ZIO.ifM(ref.getAndSet(Acquiring).map(_ == Running))(
```

Comment:
Great catches all around here. The explicit state machine makes the mechanism very clear 👍